pkg/auth/oauth2authenticator.go

```diff
@@ -0,0 +1,287 @@
+	return env, nil
+}
+
+func init() {
+	var seed int64
+	var b [8]byte
+
+	// try to use a real random value to seed the pseudo random number generator and
+	// only fall back to time based seed if this didn't work.
+	_, err := crypto_rand.Read(b[:])
+	if err != nil {
+		seed = time.Now().UnixNano() // fallback to time only if necessary
+	} else {
+		seed = int64(binary.LittleEndian.Uint64(b[:])) // based on https://stackoverflow.com/a/54491783
+	}
+	rand.Seed(seed)
+}
+
+func openBrowser(authUrl string) {
+	_ = browser.OpenURL(authUrl)
+}
+
+func shutdownServer(server *http.Server) {
+	time.Sleep(500)
+	_ = server.Shutdown(context.Background())
+}
+
+func getRedirectUri(port int) string {
+	callback := fmt.Sprintf("http://%s:%d%s", CALLBACK_HOSTNAME, port, CALLBACK_PATH)
+	return callback
+}
+
+func getOAuthConfiguration(config configuration.Configuration) *oauth2.Config {
+	appUrl := config.GetString(configuration.WEB_APP_URL)
+	tokenUrl := strings.Replace(appUrl, "app.", "id.", 1) + "/oauth2/default/v1/token"
+	tokenUrl = "https://snyk-fedramp-alpha.okta.com/oauth2/default/v1/token" // TODO HEAD-208
+	authUrl := appUrl + "/oauth/authorize"
+
+	conf := &oauth2.Config{
+		ClientID: OAUTH_CLIENT_ID,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: tokenUrl,
+			AuthURL:  authUrl,
+		},
+	}
+	return conf
+}
+
+// This method creates a code challenge as defined in https://www.rfc-editor.org/rfc/rfc7636#section-4.2
+// It accepts a code verifier and returns the challenge as a URL safe string.
+func getCodeChallenge(verifier []byte) string {
+	shasum := sha256.Sum256(verifier)
+	return base64.RawURLEncoding.EncodeToString(shasum[:])
+}
+
+// This method creates a code verifier as defined in https://www.rfc-editor.org/rfc/rfc7636#section-4.1
+// It accepts the number of bytes it shall create and returns the verifier as a byte slice of the specified length.
+func createVerifier(count int) []byte {
+	/*
+	  unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
+	   ALPHA = %x41-5A / %x61-7A
+	   DIGIT = %x30-39
+	*/
+	lut := []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
+	verifier := make([]byte, count)
+
+	// TODO is this good enough?
+	for i := range verifier {
+		index := rand.Int() % len(lut)
+		verifier[i] = lut[index]
+	}
+
+	return verifier
+}
+
+// This method extracts an oauth2.Token from the given configuration instance if available
+func GetOAuthToken(config configuration.Configuration) (*oauth2.Token, error) {
+	oauthTokenString := config.GetString(CONFIG_KEY_OAUTH_TOKEN)
+	if len(oauthTokenString) > 0 {
+		token := &oauth2.Token{}
+		err := json.Unmarshal([]byte(oauthTokenString), token)
+		if err != nil {
+			return nil, err
+		}
+		return token, nil
+	}
+	return nil, nil
+}
+
+func NewOAuth2Authenticator(config configuration.Configuration, httpClient *http.Client) Authenticator {
+	token, _ := GetOAuthToken(config)
+	oauthConfig := getOAuthConfiguration(config)
+	config.PersistInConfigFile(CONFIG_KEY_OAUTH_TOKEN)
```

Comment:
this is a side-effect - would it be possible to do that before / after the instantiation of the authenticator?